### PR TITLE
adds disable task on CR deletion

### PR DESCRIPTION
**Pull request by sp98**
added finalizer for the volumeReplication cr. When the cr is deleted, reconciler will first disable the replication and then remove the finalizer.

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

### CODE REVIEW

File: controllers/volumereplication_controller.go [hunk 1]
```diff
@@ -128,6 +129,36 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
+		if !contains(instance.GetFinalizers(), volumeReplicationFinalizer) {
+			r.Log.Info("finalizer not found for volumeReplication. Adding finalizer")
+			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, volumeReplicationFinalizer)
+			if err := r.Client.Update(context.TODO(), instance); err != nil {
+				r.Log.Error(err, "failed to update volumeReplication object with finalizer", "objectName", instance.Name, "finalizer", volumeReplicationFinalizer)
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		if contains(instance.GetFinalizers(), volumeReplicationFinalizer) {
```
Comment (Madhu-1) [inline controllers/volumereplication_controller.go:143]: general thought as we are dependent on the PVC object and we don't have control over it. what will happen if the PVC is already deleted and now the user tries to delete the volumeReplicaiton?

Comment (sp98) [reply controllers/volumereplication_controller.go:143]: with the current code, the reconcile will keep on failing at `getPVCDataSource` and volumeReplication CR deletion will be stuck.

File: controllers/volumereplication_controller.go [hunk 2]
```diff
@@ -128,6 +129,36 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	r.Log.Info("volume handle", volumeHandle)
 
+	// check if the object is being deleted
+	if instance.GetDeletionTimestamp().IsZero() {
+		if !contains(instance.GetFinalizers(), volumeReplicationFinalizer) {
+			r.Log.Info("finalizer not found for volumeReplication. Adding finalizer")
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:135]: Which volumeReplication? Which finalizer? add that info.

File: controllers/volumereplication_controller.go [hunk 3]
```diff
@@ -128,6 +129,36 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
+
+			r.Log.Info("removing finalizer")
+			// once all finalizers have been removed, the object will be deleted
+			instance.ObjectMeta.Finalizers = remove(instance.ObjectMeta.Finalizers, volumeReplicationFinalizer)
+			if err := r.Client.Update(context.TODO(), instance); err != nil {
+				r.Log.Error(err, "failed to remove finalizer from volumeReplication object", "objectName", instance.Name, "finalizer", volumeReplicationFinalizer)
+				return reconcile.Result{}, err
+			}
+		}
+		r.Log.Info("object is terminated, skipping reconciliation")
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:158]: Which object? make it clear.

File: controllers/volumereplication_controller.go [hunk 4]
```diff
@@ -128,6 +129,36 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
+		}
+	} else {
+		if contains(instance.GetFinalizers(), volumeReplicationFinalizer) {
+			failedTask, err := r.disableVolumeReplication(volumeHandle, parameters, secret)
+			if err != nil {
+				r.Log.Error(err, "task failed", "taskName", failedTask)
+				return ctrl.Result{}, err
+			}
+
+			r.Log.Info("removing finalizer")
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:150]: which finalizer?

File: controllers/volumereplication_controller.go [hunk 5]
```diff
@@ -239,3 +270,43 @@ func (r *VolumeReplicationReconciler) resyncVolume(volumeID string, parameters,
...
+// Removes a given string from a slice and returns the new slice
+func remove(slice []string, s string) (result []string) {
+	for _, item := range slice {
+		if item == s {
+			continue
+		}
+		result = append(result, item)
+	}
+	return
+}
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:312]: These are common utility functions. Keep it in a separate `controllers/strings.go` file?

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Review by umangachapagain — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**